wasm-builder: Use riscv32emac-unknown-none-polkavm.json target

jarkkojs opened the pull request:
# Description

Closes #6335.

## Integration

N/A

## Review Notes

`RuntimeTarget` is converted to return path to the custom target JSON file

jarkkojs commented:
@koute, your example command-line gave:

```
     Compiling pallet-revive-proc-macro v0.1.0 (/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk/substrate/frame/revive/proc-macro)
  error[E0432]: unresolved imports `super::key::Key`, `super::key::LazyKey`, `super::key::get`, `super::key::set`
   --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/thread_local/os.rs:1:18
    |
  1 | use super::key::{Key, LazyKey, get, set};
    |                  ^^^  ^^^^^^^  ^^^  ^^^ no `set` in `sys::thread_local::key`
    |                  |    |        |
    |                  |    |        no `get` in `sys::thread_local::key`
    |                  |    no `LazyKey` in `sys::thread_local::key`
    |                  no `Key` in `sys::thread_local::key`
    |
  note: module `crate::collections::hash::set` exists but is inaccessible
   --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/collections/hash/mod.rs:4:1
    |
  4 | pub mod set;
    | ^^^^^^^^^^^^ not accessible

  error[E0432]: unresolved imports `crate::sys::thread_local::key::LazyKey`, `crate::sys::thread_local::key::set`
   --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/thread_local/guard/key.rs:6:37
    |
  6 | use crate::sys::thread_local::key::{LazyKey, set};
    |                                     ^^^^^^^  ^^^ no `set` in `sys::thread_local::key`
    |                                     |
    |                                     no `LazyKey` in `sys::thread_local::key`
    |
  note: module `crate::collections::hash::set` exists but is inaccessible
   --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/collections/hash/mod.rs:4:1
    |
  4 | pub mod set;
    | ^^^^^^^^^^^^ not accessible

     Compiling sc-tracing-proc-macro v11.0.0 (/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk/substrate/client/tracing/proc-macro)
  error[E0425]: cannot find function `fill_bytes` in module `sys`
    --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/random.rs:67:14
     |
  67 |         sys::fill_bytes(bytes)
     |              ^^^^^^^^^^ not found in `sys`

     Compiling pallet-staking-reward-curve v11.0.0 (/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk/substrate/frame/staking/reward-curve)
     Compiling polkavm-linker v0.14.0
     Compiling sc-chain-spec-derive v11.0.0 (/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk/substrate/client/chain-spec/derive)
     Compiling tracing-gum-proc-macro v5.0.0 (/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk/polkadot/node/gum/proc-macro)
     Compiling kitchensink-runtime v3.0.0-dev (/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk/substrate/bin/node/runtime)
  error[E0277]: the trait bound `System: core::alloc::GlobalAlloc` is not satisfied
    --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/alloc/mod.rs:59:42
     |
  59 |         let new_ptr = GlobalAlloc::alloc(alloc, new_layout);
     |                       ------------------ ^^^^^ the trait `core::alloc::GlobalAlloc` is not implemented for `System`
     |                       |
     |                       required by a bound introduced by this call

  error[E0277]: the trait bound `System: core::alloc::GlobalAlloc` is not satisfied
    --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/alloc/mod.rs:63:34
     |
  63 |             GlobalAlloc::dealloc(alloc, ptr, old_layout);
     |             -------------------- ^^^^^ the trait `core::alloc::GlobalAlloc` is not implemented for `System`
     |             |
     |             required by a bound introduced by this call

  error[E0425]: cannot find function `fill_bytes` in this scope
    --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/random/mod.rs:94:5
     |
  94 |     fill_bytes(&mut buf);
     |     ^^^^^^^^^^ not found in this scope

  error[E0277]: the trait bound `System: core::alloc::GlobalAlloc` is not satisfied
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:142:47
      |
  142 |                     GlobalAlloc::alloc_zeroed(self, layout)
      |                     ------------------------- ^^^^ the trait `core::alloc::GlobalAlloc` is not implemented for `System`
      |                     |
      |                     required by a bound introduced by this call

  error[E0277]: the trait bound `System: core::alloc::GlobalAlloc` is not satisfied
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:144:40
      |
  144 |                     GlobalAlloc::alloc(self, layout)
      |                     ------------------ ^^^^ the trait `core::alloc::GlobalAlloc` is not implemented for `System`
      |                     |
      |                     required by a bound introduced by this call

  error[E0277]: the trait bound `System: core::alloc::GlobalAlloc` is not satisfied
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:178:52
      |
  178 |                 let raw_ptr = GlobalAlloc::realloc(self, ptr.as_ptr(), old_layout, new_size);
      |                               -------------------- ^^^^ the trait `core::alloc::GlobalAlloc` is not implemented for `System`
      |                               |
      |                               required by a bound introduced by this call

  error[E0277]: the trait bound `System: core::alloc::GlobalAlloc` is not satisfied
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:220:43
      |
  220 |             unsafe { GlobalAlloc::dealloc(self, ptr.as_ptr(), layout) }
      |                      -------------------- ^^^^ the trait `core::alloc::GlobalAlloc` is not implemented for `System`
      |                      |
      |                      required by a bound introduced by this call

  error[E0277]: the trait bound `System: core::alloc::GlobalAlloc` is not satisfied
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:270:52
      |
  270 |                 let raw_ptr = GlobalAlloc::realloc(self, ptr.as_ptr(), old_layout, new_size);
      |                               -------------------- ^^^^ the trait `core::alloc::GlobalAlloc` is not implemented for `System`
      |                               |
      |                               required by a bound introduced by this call

  error[E0599]: no method named `alloc` found for struct `System` in the current scope
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:401:20
      |
  132 | pub struct System;
      | ----------------- method `alloc` not found for this struct
  ...
  401 |             System.alloc(layout)
      |                    ^^^^^
      |
      = help: items from traits can only be used if the trait is implemented and in scope
      = note: the following trait defines an item `alloc`, perhaps you need to implement it:
              candidate #1: `core::alloc::GlobalAlloc`
  help: there is a method `allocate` with a similar name
      |
  401 |             System.allocate(layout)
      |                    ~~~~~~~~

  error[E0599]: no method named `dealloc` found for struct `System` in the current scope
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:409:25
      |
  132 | pub struct System;
      | ----------------- method `dealloc` not found for this struct
  ...
  409 |         unsafe { System.dealloc(ptr, Layout::from_size_align_unchecked(size, align)) }
      |                         ^^^^^^^
      |
      = help: items from traits can only be used if the trait is implemented and in scope
      = note: the following trait defines an item `dealloc`, perhaps you need to implement it:
              candidate #1: `core::alloc::GlobalAlloc`
  help: there is a method `deallocate` with a similar name
      |
  409 |         unsafe { System.deallocate(ptr, Layout::from_size_align_unchecked(size, align)) }
      |                         ~~~~~~~~~~

  error[E0599]: no method named `realloc` found for struct `System` in the current scope
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:423:20
      |
  132 | pub struct System;
      | ----------------- method `realloc` not found for this struct
  ...
  423 |             System.realloc(ptr, old_layout, new_size)
      |                    ^^^^^^^ method not found in `System`
      |
      = help: items from traits can only be used if the trait is implemented and in scope
      = note: the following trait defines an item `realloc`, perhaps you need to implement it:
              candidate #1: `core::alloc::GlobalAlloc`

  error[E0599]: no method named `alloc_zeroed` found for struct `System` in the current scope
     --> /Users/jarkko/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/alloc.rs:433:20
      |
  132 | pub struct System;
      | ----------------- method `alloc_zeroed` not found for this struct
  ...
  433 |             System.alloc_zeroed(layout)
      |                    ^^^^^^^^^^^^
      |
      = help: items from traits can only be used if the trait is implemented and in scope
      = note: the following trait defines an item `alloc_zeroed`, perhaps you need to implement it:
              candidate #1: `core::alloc::GlobalAlloc`
  help: there is a method `allocate_zeroed` with a similar name
      |
  433 |             System.allocate_zeroed(layout)
      |                    ~~~~~~~~~~~~~~~

  Some errors have detailed explanations: E0277, E0425, E0432, E0599.
  For more information about an error, try `rustc --explain E0277`.
  error: could not compile `std` (lib) due to 15 previous errors
  warning: build failed, waiting for other jobs to finish...
( SUBSTRATE_RUNTIME_TARGET=riscv SUBSTRATE_ENABLE_POLKAVM=1 RUST_LOG= cargo  )  251.53s user 124.97s system 352% cpu 1:46.68 total
```

jarkkojs commented:
How do I pass `check-labels`?

koute commented:
> How do I pass `check-labels`?

If you click on the job it tells you why it fails:

> You need to include some of the T* label(s)

AFAIK, [here](https://github.com/paritytech/labels/tree/main/ruled_labels) is the repo with the rules, and [here](https://paritytech.github.io/labels/doc_polkadot-sdk.html) are som docs.

jarkkojs commented:
depends on https://github.com/paritytech/polkavm/pull/213

jarkkojs commented:
> So the issue here is compiling PolkaVM itself to `risc32emac`? Since the `kitchensink` runtime uses PolkaVM to run contracts?

I th

> > How do I pass `check-labels`?
> 
> If you click on the job it tells you why it fails:
> 
> > You need to include some of the T* label(s)
> 
> AFAIK, [here](https://github.com/paritytech/labels/tree/main/ruled_labels) is the repo with the rules, and [here](https://paritytech.github.io/labels/doc_polkadot-sdk.html) are some docs.

Aah so labels are labels in Github? OK cool I thought I has something missing in the commit itself.

athei commented:
Yes it checks that you have all the required github labels

jarkkojs commented:
@koute looking pretty good but i don't know what the remaining error messages even mean :-)

athei commented:
They seem all to be unrelated to your changes. The CI is notoriously flaky. Merging master and praying is what you do in these cases.

athei commented on substrate/utils/wasm-builder/src/wasm_project.rs:
```diff
@@ -910,7 +919,10 @@ fn build_bloaty_blob(
 	if crate::build_std_required() {
```
We always need to build core,allocation for RISCV since it isn't shipped with rustup. But we guard it behind this conditional.

jarkkojs replied:
So, what I need to do here?

koute replied:
The way it currently is is fine. :P The `build-std` will be always enabled by default for RISC-V anyway.

jarkkojs replied:
OK resolving this one then!

athei commented on substrate/utils/wasm-builder/src/wasm_project.rs:
Can't mark this because it wasn't changes but in this file you can also remove the extra parameters added for RISC-V like: `-C link-arg=--unique` as they are now part of the target.

jarkkojs replied:
Looking at:

```
			rustflags.push_str("-C target-feature=+lui-addi-fusion -C relocation-model=pie -C link-arg=--emit-relocs -C link-arg=--unique ");
```

Should I remove all? Just checking.

koute replied:
Yes, these should be removed.

jarkkojs replied:
ah i forgot this! sorry I'll remove it

koute commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -411,7 +380,7 @@ fn get_bool_environment_variable(name: &str) -> Option<bool> {
 /// Returns whether we need to also compile the standard library when compiling the runtime.
 fn build_std_required() -> bool {
-	let default = runtime_target() == RuntimeTarget::Wasm;
+	let default = RuntimeTarget::new() != None;
 
 	crate::get_bool_environment_variable(crate::WASM_BUILD_STD).unwrap_or(default)
```
```suggestion
	crate::get_bool_environment_variable(crate::WASM_BUILD_STD).unwrap_or(true)
```

The default check here is pointless; if `runtime_target` returns `None` then the build will return an error and exit anyway, so effectively `default` is always going to be `true` here.

koute commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -423,36 +392,35 @@ enum RuntimeTarget {
+		if self == RuntimeTarget::Wasm {
+			return "wasm32-unknown-unknown".to_string();
 		}
+		let path: &'static str = env!("CARGO_MANIFEST_DIR");
+		format!("{path}/riscv32emac-unknown-none-polkavm.json")
```
Use `Path::join` instead of `format!`.

jarkkojs replied:
I'll use match also here

koute commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -423,36 +392,35 @@ enum RuntimeTarget {
-			RuntimeTarget::Riscv => "rbuild",
+	/// Figures out the target parameter value for rustc.
+	fn parameter(self) -> String {
+		if self == RuntimeTarget::Wasm {
```
Use a `match` instead of `==` to guarantee every variant is exhaustively handled.

koute commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -423,36 +392,35 @@ enum RuntimeTarget {
-		std::process::exit(1);
+	/// Figures out the target directory name used by cargo.
+	fn directory(self) -> &'static str {
+		if self == RuntimeTarget::Wasm {
```
Use `match`.

koute commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -423,36 +392,35 @@ enum RuntimeTarget {
-			RuntimeTarget::Wasm => "wbuild",
-			RuntimeTarget::Riscv => "rbuild",
+	/// Figures out the target parameter value for rustc.
+	fn parameter(self) -> String {
```
`parameter` -> `rustc_target_arg` maybe?

jarkkojs replied:
I reverted the old name.

athei commented:
I am confused that this even work. The `polkavm-linker` and `polkavm-derive` dependencies is still set to `0.9.3`. Shouldn't it be bumped in this PR, too? 
https://github.com/paritytech/polkadot-sdk/blob/a1af8ed63668fc4a553776e381a12d165e7462f8/Cargo.toml#L1093-L1094

jarkkojs commented:
> I am confused that this even works. The `polkavm-linker` and `polkavm-derive` dependencies is still set to `0.9.3`. Shouldn't it be bumped in this PR, too?
> 
> https://github.com/paritytech/polkadot-sdk/blob/a1af8ed63668fc4a553776e381a12d165e7462f8/Cargo.toml#L1093-L1094

I have no idea, @koute ?

koute commented:
https://github.com/paritytech/polkadot-sdk/actions/runs/11835548683/job/32978647159?pr=6419#step:5:988

Looks like the build runtime tests are failing, so yes, the derive and the linker crates should also be bumped.

jarkkojs commented:
> https://github.com/paritytech/polkadot-sdk/actions/runs/11835548683/job/32978647159?pr=6419#step:5:988
> 
> Looks like the build runtime tests are failing, so yes, the derive and the linker crates should also be bumped.

@koute, @athei 

I posted this on Tuesday:

https://github.com/paritytech/polkadot-sdk/pull/6419#issuecomment-2471537155

athei commented:
Answered here: https://github.com/paritytech/polkadot-sdk/pull/6419#issuecomment-2474755606

athei commented:
Please don't merge without addressing:
https://github.com/paritytech/polkadot-sdk/pull/6419#issuecomment-2476075272

jarkkojs commented:
![image](https://github.com/user-attachments/assets/f79cabf4-be1c-4424-b4ce-32e10a6ae4b3)

Doing a local run.

jarkkojs commented:
So if I do `act --container-architecture linux/amd64 -W .github/workflows/build-misc.yml`, I get:

```
INFO[0000] Using docker host 'unix:///var/run/docker.sock', and daemon socket 'unix:///var/run/docker.sock'
[preflight/Preflight/preflight] 🚀  Start image=catthehacker/ubuntu:act-latest
[preflight/Preflight/preflight]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform=linux/amd64 username= forcePull=true
[preflight/Preflight/preflight]   🐳  docker create image=catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[] network="host"
[preflight/Preflight/preflight]   🐳  docker run image=catthehacker/ubuntu:act-latest platform=linux/amd64 entrypoint=["tail" "-f" "/dev/null"] cmd=[] network="host"
[preflight/Preflight/preflight]   🐳  docker exec cmd=[node --no-warnings -e console.log(process.execPath)] user= workdir=
[preflight/Preflight/preflight]   ☁  git clone 'https://github.com/tj-actions/changed-files' # ref=v45
[preflight/Preflight/preflight] ⭐ Run Main actions/checkout@v4
[preflight/Preflight/preflight]   🐳  docker cp src=/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk/. dst=/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk
[preflight/Preflight/preflight]   ✅  Success - Main actions/checkout@v4
[preflight/Preflight/preflight] ⭐ Run Main Current file
[preflight/Preflight/preflight]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/current_file.sh] user= workdir=
[preflight/Preflight/preflight]   ✅  Success - Main Current file
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflowFile=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentActionDir=
[preflight/Preflight/preflight] ⭐ Run Main Set changes
[preflight/Preflight/preflight]   🐳  docker cp src=/Users/jarkko/.cache/act/tj-actions-changed-files@v45/ dst=/var/run/act/actions/tj-actions-changed-files@v45/
[preflight/Preflight/preflight]   🐳  docker exec cmd=[/opt/acttoolcache/node/18.20.5/x64/bin/node /var/run/act/actions/tj-actions-changed-files@v45/dist/index.js] user= workdir=
[preflight/Preflight/preflight]   ❓  ::group::changed-files
[preflight/Preflight/preflight]   💬  ::debug::Env: {%0A  "GITHUB_REF_NAME": "wasm-builder-riscv32emac",%0A  "GITHUB_REF": "refs/heads/wasm-builder-riscv32emac",%0A  "GITHUB_WORKSPACE": "/Users/jarkko/work/github.com/jarkkojs/polkadot-sdk"%0A}
[preflight/Preflight/preflight]   💬  ::debug::Inputs: {%0A  "files": "",%0A  "filesSeparator": "\n",%0A  "filesFromSourceFile": "",%0A  "filesFromSourceFileSeparator": "\n",%0A  "filesYaml": "rust:\n  - '**/*'\n  - '!.github/**/*'\n  - '!prdoc/**/*'\n  - '!docs/**/*'\ncurrentWorkflow:\n  - ''\n  - '.github/workflows/reusable-preflight.yml'",%0A  "filesYamlFromSourceFile": "",%0A  "filesYamlFromSourceFileSeparator": "\n",%0A  "filesIgnore": "",%0A  "filesIgnoreSeparator": "\n",%0A  "filesIgnoreFromSourceFile": "",%0A  "filesIgnoreFromSourceFileSeparator": "\n",%0A  "filesIgnoreYaml": "",%0A  "filesIgnoreYamlFromSourceFile": "",%0A  "filesIgnoreYamlFromSourceFileSeparator": "\n",%0A  "failOnInitialDiffError": false,%0A  "failOnSubmoduleDiffError": false,%0A  "separator": " ",%0A  "sha": "",%0A  "baseSha": "",%0A  "since": "",%0A  "until": "",%0A  "path": ".",%0A  "quotepath": true,%0A  "diffRelative": true,%0A  "sinceLastRemoteCommit": false,%0A  "recoverDeletedFiles": false,%0A  "recoverDeletedFilesToDestination": "",%0A  "recoverFiles": "",%0A  "recoverFilesSeparator": "\n",%0A  "recoverFilesIgnore": "",%0A  "recoverFilesIgnoreSeparator": "\n",%0A  "includeAllOldNewRenamedFiles": false,%0A  "oldNewSeparator": ",",%0A  "oldNewFilesSeparator": " ",%0A  "skipInitialFetch": false,%0A  "fetchAdditionalSubmoduleHistory": false,%0A  "dirNamesDeletedFilesIncludeOnlyDeletedDirs": false,%0A  "excludeSubmodules": false,%0A  "usePosixPathSeparator": false,%0A  "tagsPattern": "*",%0A  "tagsIgnorePattern": "",%0A  "dirNames": false,%0A  "dirNamesExcludeCurrentDir": false,%0A  "dirNamesIncludeFiles": "",%0A  "dirNamesIncludeFilesSeparator": "\n",%0A  "json": false,%0A  "escapeJson": true,%0A  "safeOutput": true,%0A  "writeOutputFiles": false,%0A  "outputDir": ".github/outputs",%0A  "outputRenamedFilesAsDeletedAndAdded": false,%0A  "token": "",%0A  "apiUrl": "https://api.github.com",%0A  "negationPatternsFirst": false,%0A  "useRestApi": false,%0A  "fetchDepth": 25,%0A  "fetchMissingHistoryMaxRetries": 20%0A}
[preflight/Preflight/preflight]   💬  ::debug::Working directory: /Users/jarkko/work/github.com/jarkkojs/polkadot-sdk
[preflight/Preflight/preflight]   💬  ::debug::Has git directory: true
[preflight/Preflight/preflight]   💬  ::debug::Input file patterns: %0A
[preflight/Preflight/preflight]   💬  ::debug::File patterns: 
[preflight/Preflight/preflight]   💬  ::debug::Yaml file patterns: {"rust":["**/*","!.github/**/*","!prdoc/**/*","!docs/**/*"],"currentWorkflow":[".github/workflows/reusable-preflight.yml"]}
[preflight/Preflight/preflight]   | Using local .git directory
[preflight/Preflight/preflight]   | Running on a push event...
[preflight/Preflight/preflight]   💬  ::debug::Getting current SHA...
[preflight/Preflight/preflight]   💬  ::debug::Current SHA: 8fcdcbbb0250f469d04d18653ac5a88db90a73df
[preflight/Preflight/preflight]   💬  ::debug::Getting previous SHA...
[preflight/Preflight/preflight]   💬  ::debug::Getting previous SHA for last remote commit...
[preflight/Preflight/preflight]   💬  ::debug::Previous SHA: e7334e198512d2416bbaaabb5e8ebc02f7ca0d86
[preflight/Preflight/preflight]   💬  ::debug::Target branch: wasm-builder-riscv32emac
[preflight/Preflight/preflight]   💬  ::debug::Current branch: wasm-builder-riscv32emac
[preflight/Preflight/preflight]   | Retrieving changes between e7334e198512d2416bbaaabb5e8ebc02f7ca0d86 (wasm-builder-riscv32emac) → 8fcdcbbb0250f469d04d18653ac5a88db90a73df (wasm-builder-riscv32emac)
[preflight/Preflight/preflight]   💬  ::debug::All diff files: {"A":[],"C":[],"D":[],"M":["substrate/utils/wasm-builder/src/wasm_project.rs"],"R":[],"T":[],"U":[],"X":[]}
[preflight/Preflight/preflight]   | All Done!
[preflight/Preflight/preflight]   ❓  ::endgroup::
[preflight/Preflight/preflight]   ❓  ::group::changed-files-yaml-rust
[preflight/Preflight/preflight]   💬  ::debug::All filtered diff files for rust: {"A":[],"C":[],"D":[],"M":["substrate/utils/wasm-builder/src/wasm_project.rs"],"R":[],"T":[],"U":[],"X":[]}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Added files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Copied files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Modified files: {"paths":"substrate/utils/wasm-builder/src/wasm_project.rs","count":"1"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Renamed files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Type changed files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Unmerged files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Unknown files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All changed and modified files: {"paths":"substrate/utils/wasm-builder/src/wasm_project.rs","count":"1"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All changed files: {"paths":"substrate/utils/wasm-builder/src/wasm_project.rs","count":"1"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All other changed files: {"paths":"substrate/utils/wasm-builder/src/wasm_project.rs","count":"1"}
[preflight/Preflight/preflight]   💬  ::debug::other_changed_files: []
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All modified files: {"paths":"substrate/utils/wasm-builder/src/wasm_project.rs","count":"1"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::other_modified_files: []
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Deleted files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::other_deleted_files: []
[preflight/Preflight/preflight]   | All Done!
[preflight/Preflight/preflight]   ❓  ::endgroup::
[preflight/Preflight/preflight]   ❓  ::group::changed-files-yaml-currentWorkflow
[preflight/Preflight/preflight]   💬  ::debug::All filtered diff files for currentWorkflow: {"A":[],"C":[],"D":[],"M":[],"R":[],"T":[],"U":[],"X":[]}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Added files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Copied files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Modified files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Renamed files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Type changed files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Unmerged files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Unknown files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All changed and modified files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All changed files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All other changed files: {"paths":"substrate/utils/wasm-builder/src/wasm_project.rs","count":"1"}
[preflight/Preflight/preflight]   💬  ::debug::other_changed_files: ["substrate/utils/wasm-builder/src/wasm_project.rs"]
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::All modified files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::other_modified_files: ["substrate/utils/wasm-builder/src/wasm_project.rs"]
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::Deleted files: {"paths":"","count":"0"}
[preflight/Preflight/preflight]   💬  ::debug::Dir names include file patterns: []
[preflight/Preflight/preflight]   💬  ::debug::other_deleted_files: []
[preflight/Preflight/preflight]   | All Done!
[preflight/Preflight/preflight]   ❓  ::endgroup::
[preflight/Preflight/preflight]   💬  ::debug::modified_keys: ["rust"]
[preflight/Preflight/preflight]   💬  ::debug::changed_keys: ["rust"]
[preflight/Preflight/preflight]   ✅  Success - Main Set changes
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_any_modified=true
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_modified_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_unmerged_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_modified_files_count=1
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_deleted_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_type_changed_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_all_changed_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_any_modified=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_type_changed_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_any_changed=true
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_other_changed_files_count=1
[preflight/Preflight/preflight]   ⚙  ::set-output:: modified_keys=rust
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_unknown_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_only_changed=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_all_modified_files_count=1
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_only_modified=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_deleted_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_unknown_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_other_modified_files=substrate/utils/wasm-builder/src/wasm_project.rs
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_renamed_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_all_changed_and_modified_files=substrate/utils/wasm-builder/src/wasm_project.rs
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_unmerged_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_any_deleted=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_unmerged_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_only_deleted=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_unknown_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_any_deleted=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_added_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_unknown_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_copied_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_type_changed_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_other_modified_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_deleted_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_unmerged_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_only_modified=true
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_other_changed_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_all_modified_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_other_deleted_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_type_changed_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_any_changed=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_other_changed_files=substrate/utils/wasm-builder/src/wasm_project.rs
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_only_deleted=false
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_modified_files=substrate/utils/wasm-builder/src/wasm_project.rs
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_other_modified_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_modified_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_other_deleted_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_all_changed_files_count=1
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_all_modified_files=substrate/utils/wasm-builder/src/wasm_project.rs
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_all_changed_and_modified_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_all_changed_and_modified_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_other_modified_files_count=1
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_only_changed=true
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_other_deleted_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_renamed_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_deleted_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_other_deleted_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: changed_keys=rust
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_added_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_other_changed_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_renamed_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_all_modified_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_copied_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_all_changed_and_modified_files_count=1
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_all_changed_files=substrate/utils/wasm-builder/src/wasm_project.rs
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_added_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_copied_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_copied_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: currentWorkflow_all_changed_files=
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_added_files_count=0
[preflight/Preflight/preflight]   ⚙  ::set-output:: rust_renamed_files=
[preflight/Preflight/preflight] ⭐ Run Main Set image
[preflight/Preflight/preflight]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/set_image.sh] user= workdir=
[preflight/Preflight/preflight]   ✅  Success - Main Set image
[preflight/Preflight/preflight]   ⚙  ::set-output:: IMAGE="docker.io/paritytech/ci-unified:bullseye-1.81.0-2024-09-11-v202409111034"
[preflight/Preflight/preflight] ⭐ Run Main Set runner
[preflight/Preflight/preflight]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/set_runner.sh] user= workdir=
[preflight/Preflight/preflight]   ✅  Success - Main Set runner
[preflight/Preflight/preflight]   ⚙  ::set-output:: RUNNER_MACOS=parity-macos
[preflight/Preflight/preflight]   ⚙  ::set-output:: RUNNER=parity-large
[preflight/Preflight/preflight]   ⚙  ::set-output:: RUNNER_OLDLINUX=parity-oldlinux
[preflight/Preflight/preflight]   ⚙  ::set-output:: RUNNER_DEFAULT=parity-default
[preflight/Preflight/preflight]   ⚙  ::set-output:: RUNNER_WEIGHTS=parity-weights
[preflight/Preflight/preflight]   ⚙  ::set-output:: RUNNER_BENCHMARK=parity-benchmark
[preflight/Preflight/preflight] ⭐ Run Main Set vars
[preflight/Preflight/preflight]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/set_vars.sh] user= workdir=
[preflight/Preflight/preflight]   ✅  Success - Main Set vars
[preflight/Preflight/preflight]   ⚙  ::set-output:: REF_SLUG=wasm-builder-riscv32emac
[preflight/Preflight/preflight]   ⚙  ::set-output:: SOURCE_REF_SLUG=wasm-builder-riscv32emac
[preflight/Preflight/preflight]   ⚙  ::set-output:: COMMIT_SHA=8fcdcbbb0250f469d04d18653ac5a88db90a73df
[preflight/Preflight/preflight]   ⚙  ::set-output:: COMMIT_SHA_SHORT=8fcdcbbb
[preflight/Preflight/preflight] ⭐ Run Main log
[preflight/Preflight/preflight]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/6.sh] user= workdir=
[preflight/Preflight/preflight]   | workflow file: 
[preflight/Preflight/preflight]   | Modified: rust
[preflight/Preflight/preflight]   ✅  Success - Main log
[preflight/Preflight/preflight] Cleaning up container for job preflight
[preflight/Preflight/preflight] 🏁  Job succeeded
[preflight/Preflight/ci-versions] 🚀  Start image="docker.io/paritytech/ci-unified:bullseye-1.81.0-2024-09-11-v202409111034"
[preflight/Preflight/ci-versions]   🐳  docker pull image="docker.io/paritytech/ci-unified:bullseye-1.81.0-2024-09-11-v202409111034" platform=linux/amd64 username= forcePull=true
[preflight/Preflight/ci-versions] invalid reference format
```

jarkkojs commented:
So can we merge this?

athei commented:
No. The CI is failing.

athei commented on Cargo.toml:
```diff
@@ -1090,8 +1090,8 @@ polkadot-test-client = { path = "polkadot/node/test/client" }
-polkavm-derive = "0.9.1"
-polkavm-linker = "0.9.2"
+polkavm-derive = "0.16.0"
+polkavm-linker = "0.16.0"
```
Even though this also resolved to `0.16.1` we still want to make sure that the buggy version isn't a legal configuration.

```suggestion
polkavm-linker = "0.16.1"
```

athei commented on substrate/utils/wasm-builder/riscv32emac-unknown-none-polkavm.json:
You can delete this file and replace by calling this:
https://docs.rs/polkavm-linker/latest/polkavm_linker/fn.target_json_32_path.html

jarkkojs replied:
https://github.com/paritytech/polkadot-sdk/pull/6419/commits/8104f3007aa48a6660ed7b2e9b624987960e509e

athei commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -411,9 +380,7 @@ fn get_bool_environment_variable(name: &str) -> Option<bool> {
-	let default = runtime_target() == RuntimeTarget::Wasm;
-
-	crate::get_bool_environment_variable(crate::WASM_BUILD_STD).unwrap_or(default)
+	crate::get_bool_environment_variable(crate::WASM_BUILD_STD).unwrap_or(true)
```
Let's add a comment here why.

```suggestion
    // We always rebuild std except if specifically disabled. For Wasm this is necessary because
    // the prebuilt std is compiled with unsupported features. For RISC-V there is no prebuilt binaries
    // available in rustup.
	crate::get_bool_environment_variable(crate::WASM_BUILD_STD).unwrap_or(true)
```

athei commented on substrate/utils/wasm-builder/src/wasm_project.rs:
```diff
@@ -109,6 +109,15 @@ fn crate_metadata(cargo_manifest: &Path) -> Metadata {
 
+/// Keep the build directories separate so that when switching between the
+/// targets we won't trigger unnecessary rebuilds.
+fn build_subdirectory(target: RuntimeTarget) -> &'static str {
```
Why not keep it as inherent function?

jarkkojs replied:
It's local to `wasm_project.rs`. Let's reconsider if this ever changes.

athei commented on substrate/utils/wasm-builder/src/wasm_project.rs:
```diff
@@ -841,9 +850,7 @@ fn build_bloaty_blob(
-		RuntimeTarget::Riscv => {
-			rustflags.push_str("-C target-feature=+lui-addi-fusion -C relocation-model=pie -C link-arg=--emit-relocs -C link-arg=--unique ");
-		},
+		RuntimeTarget::Riscv => (),
```
Can we move the target features to be an inherent function on `RuntimeTarget`?

jarkkojs replied:
Same as earlier. It's local to `wasm_project.rs`. Let's reconsider if this ever changes.

jarkkojs replied:
Your comment about build-std made sense, given that it has 2x sites, and improves validation in the other site. I'd set this as a common sense threshold what to merge to `RustTarget`.

athei commented on substrate/utils/wasm-builder/src/wasm_project.rs:
```diff
@@ -910,7 +917,10 @@ fn build_bloaty_blob(
 		// Unfortunately this is still a nightly-only flag, but FWIW it is pretty widely used
 		// so it's unlikely to break without a replacement.
-		build_cmd.arg("-Z").arg("build-std");
+		match target {
+			RuntimeTarget::Wasm => build_cmd.arg("-Z").arg("build-std"),
+			RuntimeTarget::Riscv => build_cmd.arg("-Z").arg("build-std=core,alloc"),
+		};
```
Let's make `build_std_required()` return an `Option<&str>` with the required `-Zbuild-std=*` string.

jarkkojs replied:
https://github.com/paritytech/polkadot-sdk/pull/6419/commits/4dd8d28e8b157f76f0cf74b0dc2b7aabfcc79c12

athei reviewed (commented):
Looks good. Just asking for a little bit of refactoring.

jarkkojs commented:
@koute possible to approve this?

koute commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -409,50 +378,58 @@ fn get_bool_environment_variable(name: &str) -> Option<bool> {
 impl RuntimeTarget {
-	fn rustc_target(self) -> &'static str {
+	/// Creates a new instance.
+	fn new() -> Option<Self> {
+		let Some(value) = env::var_os(RUNTIME_TARGET) else {
+			return Some(Self::Wasm);
+		};
+		if value == "wasm" {
+			Some(Self::Wasm)
+		} else if value == "riscv" {
+			Some(Self::Riscv)
+		} else {
+			None
+		}
+	}
```
This should either return a `Result`, or it should do what it was originally doing (a warning + exit). `None` should almost *never* be used to signal an error condition.

jarkkojs replied:
I'll change this to warning + exit. Changing that semantics is out of scope for the PR.

koute commented on substrate/utils/wasm-builder/src/prerequisites.rs:
```diff
@@ -200,7 +200,16 @@ fn check_wasm_toolchain_installed(
 	let version = dummy_crate.get_rustc_version();
-	if crate::build_std_required() {
+
+	// Validate (as a sanity-check) that the target is in-fact `RuntimeTarget::Wasm`.
+	let Some(target) = RuntimeTarget::new() else {
+		return Err(colorize_error_message("Cannot compile the WASM runtime: invalid target"));
+	};
+	if target != RuntimeTarget::Wasm {
+		return Err(colorize_error_message("Cannot compile the WASM runtime: wrong target"));
+	}
```
If you insist on checking this then this should be an assert, as the `check_wasm_toolchain_installed` should only be called when targeting WASM.

jarkkojs replied:
I'll change these to asserts. I recall faintly that this was to address a review comment but could not find a reference.

jarkkojs commented on substrate/utils/wasm-builder/riscv32emac-unknown-none-polkavm.json:
```diff
@@ -0,0 +1,26 @@
+    ]
+  },
+  "env": "polkavm"
+}
```
https://github.com/paritytech/polkadot-sdk/pull/6419/commits/3153d5ab0ae79dc64bcc611735b40340686daad4

jarkkojs replied:
https://github.com/paritytech/polkadot-sdk/pull/6419/commits/3153d5ab0ae79dc64bcc611735b40340686daad4

athei commented:
> > Review required! Latest push from author must always be reviewed
> 
> @athei @koute was just a merge

This is because you opened the PR from your fork AFAIK. Create a branch on this repo to avoid this.

athei commented:
> > > > Review required! Latest push from author must always be reviewed
> > > 
> > > 
> > > @athei @koute was just a merge
> > 
> > 
> > This is because you opened the PR from your fork AFAIK. Create a branch on this repo to avoid this.
> 
> Hmm...I might not fully understand what ur saying. I.e. I have branch in my fork and I opened PR. Isn't this the normal procedure.
> 
> Please fix my process and I will follow anything i.e. not challenging status quo ;-

What I am saying is: Don't fork this repository. You can create a branch directly on this repository since you are part of the Parity organization.

jarkkojs commented:
> > > > > Review required! Latest push from author must always be reviewed
> > > > 
> > > > 
> > > > @athei @koute was just a merge
> > > 
> > > 
> > > This is because you opened the PR from your fork AFAIK. Create a branch on this repo to avoid this.
> > 
> > 
> > Hmm...I might not fully understand what ur saying. I.e. I have branch in my fork and I opened PR. Isn't this the normal procedure.
> > Please fix my process and I will follow anything i.e. not challenging status quo ;-
> 
> What I am saying is: Don't use a fork. You can create a branch directly on this repository since you are part of the Parity organization.

OK got it. I'll use that process in future PR's.

jarkkojs commented:
> > > > > > Review required! Latest push from author must always be reviewed
> > > > > 
> > > > > 
> > > > > @athei @koute was just a merge
> > > > 
> > > > 
> > > > This is because you opened the PR from your fork AFAIK. Create a branch on this repo to avoid this.
> > > 
> > > 
> > > Hmm...I might not fully understand what ur saying. I.e. I have branch in my fork and I opened PR. Isn't this the normal procedure.
> > > Please fix my process and I will follow anything i.e. not challenging status quo ;-
> > 
> > 
> > What I am saying is: Don't use a fork. You can create a branch directly on this repository since you are part of the Parity organization.
> 
> OK got it. I'll use that process in future PR's.

OK I fixed the assert and now it is giving that same 5 errors as it has been two'ish weeks :-)

jarkkojs commented:
@athei, @koute: no other changes expect solved merge conflicts with `Cargo.lock`.

koute commented on substrate/utils/wasm-builder/src/prerequisites.rs:
```diff
@@ -196,11 +196,14 @@ fn check_wasm_toolchain_installed(
-	if crate::build_std_required() {
+
+	let target = RuntimeTarget::new();
+	assert!(target == RuntimeTarget::Wasm);
```
Should be `assert_eq!`

koute commented on substrate/utils/wasm-builder/src/prerequisites.rs:
```diff
@@ -196,11 +196,14 @@ fn check_wasm_toolchain_installed(
+
+	let target = RuntimeTarget::new();
+	assert!(target == RuntimeTarget::Wasm);
+	if target.rustc_target_build_std() != None {
```
```suggestion
	if target.rustc_target_build_std().is_some() {
```

StackOverflowExcept1on commented on substrate/utils/wasm-builder/src/lib.rs:
```diff
@@ -409,50 +378,62 @@ fn get_bool_environment_variable(name: &str) -> Option<bool> {
+		// This is a nightly-only flag.
+		let arg = match self {
+			RuntimeTarget::Wasm => "build-std",
+			RuntimeTarget::Riscv => "build-std=core,alloc",
```
@bkchr @jarkkojs What is the point of doing `-Zbuild-std` for target `riscv32emac-unknown-none-polkavm`? It increases runtime build time for non-wasm targets.

athei replied:
It is requires for both targets but for different reasons:

wasm: The rustup provided standard library is compiled with wasm extensions enabled we dont support. Hence we need to recompile it.

riscv: Rustup doesn't contain any prebuilt binaries for this target. Hence we need to build it ourselves.

jarkkojs replied:
Yeah, so I guessed, i.e. it should be official target in rustc upstream to have prebuilt binaries, right?

jarkkojs replied:
[not actually at work yet, my keyboard just writes by itself]

StackOverflowExcept1on replied:
@athei Thanks for the clarification! I think Parity could add to the pre-compiled libraries to `riscv32emac-unknown-none-polkavm` target in `rlib` format like the `wasm32-unknown-unknown` target does (located in `./lib/rustlib/wasm32-unknown-unknown/lib/libcore-192e883b194683b6.rlib`).
